Add save, restore, report cursor

mpilgrem opened the pull request:
In response to Issue #7, adds save cursor position, restore cursor position and report cursor position codes - including emulation for versions of Windows that are not ANSI-enabled.

The Windows emulation of save/restore cursor position uses the `unsafePerformIO $ newIORef` 'hack' to emulate a 'global variable' holding the saved cursor position. Saved cursor positions for different handles are held in a `Data.Map.Strict.Map`. 

The Windows emulation of report cursor position uses the Windows function `WriteConsoleInput` to inject characters into the standard input stream. The latter is provided by the Hackage package `Win32-console`.

Two examples are added to `Example.hs`, one for save/restore cursor position and one for report cursor position.

mpilgrem commented:
In respect of `Win32-console`, I thought (wrongly) that `cpp-options: -D_WIN32_WINNT=0x0600` had the same effect as `hsc2hs-options="-D_WIN32_WINNT=0x0600"`. I need to re-think that - perhaps replicating the relevant part of `Win32-console`  in `Foreign.hs`.

feuerbach commented:
Hey, sorry I wasn't paying attention to this PR. Did you have a chance to look into @joeyh's comment linked above?

mpilgrem commented:
I have looked at @joeyh's comment. I understand the request to go beyond what an ANSI terminal does when it is sent code `CSI 6n` (which is where I stopped, with this proposal) and to parse what an ANSI terminal sends to standard input.

If that is within the intended scope of the package, I could try to code a utility.

One question is where to put the utility; one utility (`hSupportsANSI`) is in module `System.Console.ANSI`, other utilities are in module `System.Console.ANSI.Codes`. It might be an opportunity to move all the utilities into a single module (say, `System.Console.ANSI.Utilities`).

feuerbach commented:
> If that is within the intended scope of the package, I could try to code a utility.

I think it is; otherwise everyone who wants to read the position would have to reimplement this function.

Also, a few other things regarding the patch:

1. Could you document the new functions? They don't need a whole lot of documentation; just a mention that save and restore are meant to be used as a pair, and what `reportCursorPosition` does. Maybe some minimal examples of how to use them. I am not sure about the best way to avoid duplicating the docs between UNIX and Windows. Maybe add a common documentation blob in the "Saving, restoring and reporting cursor position" section?
1. How about enabling hsc2hs in `System.Console.ANSI.Windows.Foreign` and replacing hardcoded peek/poke offsets with `#peek` and `#poke` macros? I think it would be more readable and maintainable.
1. There's a whitespace error in the cabal file, see `git show --check 10c093b`.
1. It need rebasing on top of master.

> One question is where to put the utility; one utility (hSupportsANSI) is in module System.Console.ANSI, other utilities are in module System.Console.ANSI.Codes. It might be an opportunity to move all the utilities into a single module (say, System.Console.ANSI.Util).

By utilities in `System.Console.ANSI.Codes`, do you mean `colorToCode`, `csi`, and `sgrToCode`? These seem like functions exposed just in case, and a typical user won't need them (correct me if I'm wrong). If that's the case, I'd prefer them stay there. OTOH, `hSupportsANSI` and the function we're discussing are ones that are intended to be called by a typical users, so they belong in the main module.

mpilgrem commented:
Covering the items above: 1. I have added documentation. 2. I'll have to read up on `hsc2hs`. It is a while ago, but my recollection is that I did try it in the past but could not get it to work. For now, I've left the code as it is, as I know it works. 3. I've changed my editor settings to automatically trim whitespace on save, so this should be fixed. 4. I have rebased. 5. Can I add the utility to parse the result as a separate pull request/commit?

feuerbach commented:
> I'll have to read up on hsc2hs. It is a while ago, but my recollection is that I did try it in the past but could not get it to work. For now, I've left the code as it is, as I know it works.

Ok, fair enough. Let me know if you need help with hsc2hs — I've been dealing with it a lot lately :)

> Can I add the utility to parse the result as a separate pull request/commit?

Sure.
